cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i

```diff
@@ -1202,79 +1235,69 @@ Status KNNClassificationTrainBatchKernel<algorithmFpType, training::defaultDense
     if (status.ok())
     {
         status = [&]() -> Status {
-            int result           = 0;
-            bool isNeedToReindex = false;
-            localTLS.reduce([=, &isNeedToReindex](Local * ptr) -> void {
-                if (ptr && ptr->extraKDTreeNodes)
+            int result = 0;
+
+            size_t actualNodeCount = lastNodeIndex;
+            localTLS.reduce([=, &actualNodeCount](Local * ptr) -> void {
+                if (ptr)
                 {
-                    isNeedToReindex = true;
+                    actualNodeCount += ptr->nodeIndex - firstNodeIndex[ptr->threadIndex];
```

Comment:
Would be interesting to see how this change impacts the performance.
But anyway, the change is justified as it fixes the crashes.